From 5a061ef405f71d62d6484f40437e6b1fabbb431b Mon Sep 17 00:00:00 2001
From: Robert Kovacsics <robert.kovacsics@carallon.com>
Date: Fri, 25 Oct 2024 12:25:20 +0100
Subject: [PATCH 05/11] Patch st-true-color.patch

---
 st.c    | 1 +
 st.info | 1 +
 2 files changed, 2 insertions(+)

diff --git a/st.c b/st.c
index 8e57991..291859e 100644
--- a/st.c
+++ b/st.c
@@ -697,6 +697,7 @@ execsh(char *cmd, char **args)
 	setenv("SHELL", sh, 1);
 	setenv("HOME", pw->pw_dir, 1);
 	setenv("TERM", termname, 1);
+	setenv("COLORTERM", "truecolor", 1);
 
 	signal(SIGCHLD, SIG_DFL);
 	signal(SIGHUP, SIG_DFL);
diff --git a/st.info b/st.info
index efab2cf..7c81d67 100644
--- a/st.info
+++ b/st.info
@@ -191,6 +191,7 @@ st-mono| simpleterm monocolor,
 # disabled rep for now: causes some issues with older ncurses versions.
 #	rep=%p1%c\E[%p2%{1}%-%db,
 # tmux extensions, see TERMINFO EXTENSIONS in tmux(1)
+	RGB,
 	Tc,
 	Ms=\E]52;%p1%s;%p2%s\007,
 	Se=\E[2 q,
-- 
2.50.1

