From cffe3abe0d3be3ccdff5e52738db9ea4aeb64b11 Mon Sep 17 00:00:00 2001
From: Kovacsics Robert <kovirobi@gmail.com>
Date: Fri, 16 Jan 2026 20:58:14 +0000
Subject: [PATCH 2/3] Exit code when attaching to dead session

---
 abduco.c | 24 +++++++++++++++++++-----
 client.c |  7 -------
 server.c | 17 ++++++++++++++++-
 3 files changed, 35 insertions(+), 13 deletions(-)

diff --git a/abduco.c b/abduco.c
index d1dbd57..aa1a29f 100644
--- a/abduco.c
+++ b/abduco.c
@@ -567,8 +567,6 @@ static bool attach_session(const char *name, const bool terminate) {
 		close(server.socket);
 	if ((server.socket = session_connect(name)) == -1)
 		return false;
-	if (server_set_socket_non_blocking(server.socket) == -1)
-		return false;
 
 	struct sigaction sa;
 	sa.sa_flags = 0;
@@ -578,9 +576,25 @@ static bool attach_session(const char *name, const bool terminate) {
 	sa.sa_handler = SIG_IGN;
 	sigaction(SIGPIPE, &sa, NULL);
 
-	client_setup_terminal();
-	int status = client_mainloop();
-	client_restore_terminal();
+	Packet pkt = {
+		.type = MSG_ATTACH,
+		.u.i = client.flags,
+		.len = sizeof(pkt.u.i),
+	};
+	client_send_packet(&pkt);
+
+	int status;
+	if (client_recv_packet(&pkt) && pkt.type == MSG_PID) {
+		if (server_set_socket_non_blocking(server.socket) == -1)
+			return false;
+		client_setup_terminal();
+		status = client_mainloop();
+		client_restore_terminal();
+	} else if (pkt.type == MSG_EXIT) {
+		status = pkt.u.i;
+		client_send_packet(&pkt);
+		close(server.socket);
+	}
 	if (status == -1) {
 		info("detached");
 	} else if (status == -EIO) {
diff --git a/client.c b/client.c
index 4ab447d..603c388 100644
--- a/client.c
+++ b/client.c
@@ -64,13 +64,6 @@ static int client_mainloop(void) {
 
 	bool stdin_eof = false;
 	client.need_resize = true;
-	Packet pkt = {
-		.type = MSG_ATTACH,
-		.u.i = client.flags,
-		.len = sizeof(pkt.u.i),
-	};
-	client_send_packet(&pkt);
-
 	while (server.running) {
 		fd_set fds;
 		FD_ZERO(&fds);
diff --git a/server.c b/server.c
index 9611f21..cbcf707 100644
--- a/server.c
+++ b/server.c
@@ -314,12 +314,27 @@ static void server_mainloop(void) {
 				case MSG_CONTENT:
 					server_write_pty(&client_packet);
 					break;
-				case MSG_ATTACH:
+				case MSG_ATTACH: {
+					Packet pkt;
+					if (!server.running && server.exit_status != -1) {
+						pkt = (Packet) {
+							.type = MSG_EXIT,
+							.u.i = server.exit_status,
+							.len = sizeof(pkt.u.i),
+						};
+					} else {
+						pkt = (Packet) {
+							.type = MSG_CONTENT,
+							.len = 0,
+						};
+					}
+					server_send_packet(c, &pkt);
 					c->flags = client_packet.u.i;
 					if (c->flags & CLIENT_LOWPRIORITY)
 						server_sink_client();
 					server_send_screen_buffer(c);
 					break;
+				}
 				case MSG_RESIZE: {
 						pid_t group_id;
 
-- 
2.52.0

