From 261533b8504d16cab883fe4f01d658cc92090f39 Mon Sep 17 00:00:00 2001
From: Kovacsics Robert <kovirobi@gmail.com>
Date: Fri, 16 Jan 2026 20:59:23 +0000
Subject: [PATCH 3/3] report pixel sizes to child processes

---
 abduco.c | 4 ++++
 client.c | 2 +-
 server.c | 2 ++
 3 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/abduco.c b/abduco.c
index aa1a29f..28abb65 100644
--- a/abduco.c
+++ b/abduco.c
@@ -81,6 +81,8 @@ typedef struct {
 		struct {
 			uint16_t rows;
 			uint16_t cols;
+			uint16_t xpixels;
+			uint16_t ypixels;
 		} ws;
 		uint32_t i;
 		uint64_t l;
@@ -753,6 +755,8 @@ int main(int argc, char *argv[]) {
 	if (ioctl(STDIN_FILENO, TIOCGWINSZ, &server.winsize) == -1) {
 		server.winsize.ws_col = 80;
 		server.winsize.ws_row = 25;
+		server.winsize.ws_xpixel = 640;
+		server.winsize.ws_ypixel = 480;
 	}
 
 	server.read_pty = (action == 'n');
diff --git a/client.c b/client.c
index 603c388..c4e8cdf 100644
--- a/client.c
+++ b/client.c
@@ -76,7 +76,7 @@ static int client_mainloop(void) {
 			if (ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) != -1) {
 				Packet pkt = {
 					.type = MSG_RESIZE,
-					.u = { .ws = { .rows = ws.ws_row, .cols = ws.ws_col } },
+					.u = { .ws = { .rows = ws.ws_row, .cols = ws.ws_col, .xpixels = ws.ws_xpixel, .ypixels = ws.ws_ypixel } },
 					.len = sizeof(pkt.u.ws),
 				};
 				if (client_send_packet(&pkt))
diff --git a/server.c b/server.c
index cbcf707..17b4849 100644
--- a/server.c
+++ b/server.c
@@ -344,6 +344,8 @@ static void server_mainloop(void) {
 							struct winsize ws = { 0 };
 							ws.ws_row = client_packet.u.ws.rows;
 							ws.ws_col = client_packet.u.ws.cols;
+							ws.ws_xpixel = client_packet.u.ws.xpixels;
+							ws.ws_ypixel = client_packet.u.ws.ypixels;
 							ioctl(server.pty, TIOCSWINSZ, &ws);
 						}
 
-- 
2.52.0

