From a3252f4d74f0ab505ed3cb0c6171c52d2c67a4b9 Mon Sep 17 00:00:00 2001
From: Kovacsics Robert <kovirobi@gmail.com>
Date: Sat, 3 Jan 2026 14:37:14 +0000
Subject: [PATCH] Fix pasystray compilation

---
 src/options.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/options.c b/src/options.c
index 284c27f..b35afa4 100644
--- a/src/options.c
+++ b/src/options.c
@@ -19,6 +19,7 @@
   USA.
 ***/
 
+# define _XOPEN_SOURCE	700
 #include <stdlib.h>
 #include <glib.h>
 #include <gtk/gtk.h>
-- 
2.51.2

